File: backend/app/schemas/instalacion.py
import uuid
from datetime import date, datetime
from typing import TYPE_CHECKING

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.models.instalacion import EstadoInstalacion
from app.models.cliente import TipoIdentificacion
from app.models.contrato import EstadoContrato

if TYPE_CHECKING:
    from app.schemas.plan import PlanResponse
    from app.schemas.contrato import ContratoDetailResponse


class InstalacionBase(BaseModel):
    plan_id: uuid.UUID
    fecha_programada: date
    tecnico_asignado: str | None = None
    notas: str | None = None


class InstalacionSolicitudCreate(InstalacionBase):
    temp_tipo_identificacion: TipoIdentificacion
    temp_numero_identificacion: str = Field(..., min_length=1, max_length=20)
    temp_nombre: str = Field(..., min_length=1, max_length=100)
    temp_apellido1: str | None = Field(None, max_length=100)
    temp_apellido2: str | None = Field(None, max_length=100)
    temp_razon_social: str | None = Field(None, max_length=255)
    temp_email: EmailStr | None = None
    temp_telefono: str | None = Field(None, max_length=20)
    temp_provincia: str | None = Field(None, max_length=50)
    temp_canton: str | None = Field(None, max_length=50)
    temp_distrito: str | None = Field(None, max_length=50)
    temp_direccion_exacta: str | None = None

    @field_validator("temp_telefono")
    @classmethod
    def validate_telefono(cls, v: str | None) -> str | None:
        if v is None or v == "":
            return None
        if not v.isdigit() or len(v) != 8:
            raise ValueError("Teléfono debe ser 8 dígitos")
        if v[0] not in "2345678":
            raise ValueError("Teléfono debe iniciar con dígito del 2 al 8")
        return v


class InstalacionUpdate(BaseModel):
    fecha_programada: date | None = None
    fecha_completada: date | None = None
    tecnico_asignado: str | None = None
    estado: EstadoInstalacion | None = None
    notas: str | None = None


class InstalacionResponse(BaseModel):
    id: uuid.UUID
    numero_instalacion: str
    contrato_id: uuid.UUID | None
    plan_id: uuid.UUID
    fecha_programada: date
    fecha_completada: date | None
    tecnico_asignado: str | None
    estado: EstadoInstalacion
    notas: str | None

    temp_tipo_identificacion: str | None
    temp_numero_identificacion: str | None
    temp_nombre: str | None
    temp_apellido1: str | None
    temp_apellido2: str | None
    temp_razon_social: str | None
    temp_email: str | None
    temp_telefono: str | None
    temp_provincia: str | None
    temp_canton: str | None
    temp_distrito: str | None
    temp_direccion_exacta: str | None

    pdf_solicitud_path: str | None

    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InstalacionDetailResponse(InstalacionResponse):
    plan: "PlanResponse"
    contrato: "ContratoDetailResponse | None" = None


class InstalacionActivarRequest(BaseModel):
    crear_cliente: bool = True
    cliente_id: uuid.UUID | None = None
    fecha_inicio_contrato: date
    dia_facturacion: int = Field(default=1, ge=1, le=28)
    estado_contrato: EstadoContrato = EstadoContrato.ACTIVO


# Resolve forward references
def _resolve_forward_refs():
    """Resolve forward references after all schemas are defined"""
    try:
        from app.schemas.plan import PlanResponse
        from app.schemas.contrato import ContratoDetailResponse

        InstalacionDetailResponse.model_rebuild()
    except ImportError:
        # Schemas not yet available, will be resolved later
        pass


_resolve_forward_refs()

File: frontend/src/types/instalacion.ts
import type { Cliente, TipoIdentificacion } from "./cliente";
import type { Contrato } from "./contrato";
import type { Plan } from "./plan";

export type EstadoInstalacion =
  | "solicitud"
  | "programada"
  | "en_progreso"
  | "completada"
  | "cancelada";

export interface Instalacion {
  id: string;
  numero_instalacion: string;
  contrato_id: string | null;
  plan_id: string;
  fecha_programada: string;
  fecha_completada: string | null;
  tecnico_asignado: string | null;
  estado: EstadoInstalacion;
  notas: string | null;
  motivo_cancelacion: string | null;

  temp_tipo_identificacion: TipoIdentificacion | null;
  temp_numero_identificacion: string | null;
  temp_nombre: string | null;
  temp_apellido1: string | null;
  temp_apellido2: string | null;
  temp_razon_social: string | null;
  temp_email: string | null;
  temp_telefono: string | null;
  temp_provincia: string | null;
  temp_canton: string | null;
  temp_distrito: string | null;
  temp_direccion_exacta: string | null;

  pdf_solicitud_path: string | null;

  created_at: string;
  updated_at: string;

  plan: Plan;
  contrato: Contrato | null;
}

export interface InstalacionSolicitudCreate {
  plan_id: string;
  fecha_programada: string;
  tecnico_asignado?: string;
  notas?: string;

  temp_tipo_identificacion: TipoIdentificacion;
  temp_numero_identificacion: string;
  temp_nombre: string;
  temp_apellido1?: string;
  temp_apellido2?: string;
  temp_razon_social?: string;
  temp_email?: string;
  temp_telefono?: string;
  temp_provincia?: string;
  temp_canton?: string;
  temp_distrito?: string;
  temp_direccion_exacta?: string;
}

export interface InstalacionUpdate {
  fecha_programada?: string;
  fecha_completada?: string;
  tecnico_asignado?: string;
  estado?: EstadoInstalacion;
  notas?: string;
  motivo_cancelacion?: string;
}

export interface InstalacionActivarRequest {
  crear_cliente?: boolean;
  cliente_id?: string;
  fecha_inicio_contrato: string;
  dia_facturacion?: number;
  estado_contrato?: "activo" | "suspendido" | "cancelado" | "pendiente";
  router_id: string;
  ip_asignada: string;
}
